commandapi-core/src/main/java/dev/jorel/commandapi/executors/CommandArguments.java

```diff
@@ -443,4 +445,55 @@ public <T> Optional<T> getOptionalUnchecked(String nodeName) {
 		return (Optional<T>) getOptional(nodeName);
 	}
 
+	/**
+	 * Returns an argument purely based on its CommandAPI representation. This also attempts to directly cast the argument to the type represented by {@link dev.jorel.commandapi.arguments.AbstractArgument#getPrimitiveType()}
+	 *
+	 * @param argumentType The argument instance used to create the argument
+	 * @return The argument represented by the CommandAPI argument, or null if the argument's clas cannot be cast to the type represented by {@link dev.jorel.commandapi.arguments.AbstractArgument#getPrimitiveType()}
+	 */
+	@Nullable
+	public <T> T getByArgument(AbstractArgument<T, ?, ?, ?> argumentType) {
+		Object argument = get(argumentType.getNodeName());
+		return castArgument(argument, argumentType.getPrimitiveType());
+	}
+
+	/**
+	 * Returns an argument based on its node name. This also attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param nodeName The node name of the argument
+	 * @param argumentType The class that represents the argument
+	 * @return The argument with the given node name, or null if the argument's class cannot be cast to the type represented by the given {@code argumentType} argument
+	 */
+	@Nullable
+	public <T> T getByClass(String nodeName, Class<T> argumentType) {
+		Object argument = get(nodeName);
+		return castArgument(argument, argumentType);
+	}
+
+	/**
+	 * Returns an argument based on its index. This also attempts to directly cast the argument to the type represented by the {@code argumentType} parameter.
+	 *
+	 * @param index The position of the argument
+	 * @param argumentType The class that represents the argument
+	 * @return The argument at the given index, or null if the argument's class cannot be cast to the type represented by the given {@code argumentType} argument
+	 */
+	@Nullable
+	public <T> T getByClass(int index, Class<T> argumentType) {
+		Object argument = get(index);
+		return castArgument(argument, argumentType);
+	}
+
+	private <T> T castArgument(Object argument, Class<T> argumentType) {
+		if (argument == null) {
+			return null;
+		}
+		if (!argument.getClass().equals(argumentType)) {
+			return null;
+		}
+		if (!argumentType.isAssignableFrom(argument.getClass())) {
+			return null;
+		}
```

Comment:
I might be missing a few things, but I think this doesn't work as expected - surely we want the `isAssignableFrom` check first?

Secondly, I don't think we should be returning `null` here (in the case that the classes don't match). Surely:

- We know what class the user provided for `argumentType`
- We know what class the actual `argument` is

Therefore, we can throw a runtime error stating something along the lines of "Expected <argument name or index> to be a <argument type>, but was given a <user provided class> instead".

Reply:
Interesting thoughts, as I said, I definitely want to run some tests next.
This whole thing was more or less inspired by [CommandContext#getArgument(String, Class)](https://github.com/Mojang/brigadier/blob/master/src/main/java/com/mojang/brigadier/context/CommandContext.java#L97-L111)
This is why I put the null check first but yeah, I guess we can flip the last two statements.

As for the `null`'s, I thought it was the better choice since we can have it return `null` anyway. If we  think exceptions are better, I certainly can change this.

Brigadier also uses those wrapper classes for the class check, maybe this is also something I have to look into. Again, I definitely will run some tests tomorrow, let's see what I get from those.

Reply:
Can we include support for optionals in this too, or would you prefer a separate PR for that?

Reply:
> Can we include support for optionals in this too, or would you prefer a separate PR for that?

We can do anything. If you'd like optionals, I guess I could arrange adding a method returning an optional too. And I don't think a additional PR is necessary, this is supposed to add this.

Reply:
> Can we include support for optionals in this too, or would you prefer a separate PR for that?

Methods for `Optional`'s (and default values for consistency) have been added 👍